### ISSUE TITLE
Questions for the assembly language of the project

### ISSUE THREAD

**Issue by aqjune** [snu-sf-class/swpp202001#6]
Please ask a question about the specification of the assembly language by replying to this issue.

**Comment by tlvkfn525** [comment]
reset tape access pin instruction의 기능에 대해서 좀 더 자세히 알려주실 수 있을까요?

**Comment by aqjune** [comment]
> reset tape access pin instruction의 기능에 대해서 좀 더 자세히 알려주실 수 있을까요?

Command `reset stack` moves the head to address 10240 with cost 2, and `reset heap` moves the head to address 20480 with cost 2 too.

**Comment by tlwpdus** [comment]
힙이 비어있는 상태에서
*p = malloc(4)
*q = malloc(4)
를 수행한 후에
p+4에 접근함으로써( Ex. load 4, p, 4 ) q에 있는 값을 가져올 수 있는 것인가요?
스펙에 따르면 malloc은 힙의 최소 주소에 allocation하게 되고, p+4는 실제로 allocated된 위치이므로 문제가 발생하지 않는 것 같습니다.

이와 연관해서 p-10248과 같은 식으로 스택 메모리에도 접근이 가능한지, 이런 식으로 접근한다면 cost가 어떻게 되는지도 궁금합니다.

**Comment by aqjune** [comment]
Yes, load 4, p, 8 will read the bytes at the other malloc (Note that the block allocated by malloc is 8-byte aligned).
Yes, it is allowed to access the stack area too. The cost is (the cost of moving the head to p-10248) +  2 (because p-10248 is pointing to the stack area).

**Comment by tlvkfn525** [comment]
제가 이해를 잘 못한 부분이 있는데 arg1~arg16 register가 오직 함수 인자 전달만을 위해 따로 있는 건가요, 아니면 r1~r16 register를 function call 시에 자동으로 arg register로 사용하는 건가요? call 시에 자동으로 이 값들을 저장해준다는 건 이걸 위해 별도로 load/store를 하지 않아도 된다는 뜻인가요?

**Comment by aqjune** [comment]
You can think in either way, but arg1 / arg16 are read-only registers, so any trial to write a value into argN will raise error.

**Comment by aqjune** [comment]
For example, an assembly instruction `call f 4 3 2 1` will store 4 into arg1 / 3 into arg2 / ... , and jump to the beginning of f.
After the function call is done, the original argN register values will be restored.

**Comment by kangz12345** [comment]
Function call 을 할 때, 자동으로 argument들은 argN register에 read-only로 들어가고, 일반 rN register는 저장되었다가 return 되었을때 자동으로 restore되는 것 같은데,
function내부로 들어갔을 때 rN register의 값은 원래 값이 유지되나요?
즉, 원래 rN에 있던 값을 function call 해서 어떤 function으로 들어갔을 때 그대로 보존됨이 보장되나요? 물론 function 내부에서 값을 수정하더라도 return 후에는 다시 이전 값으로 restore되겠지만요.

**Comment by aqjune** [comment]
Yes, they will be preserved.

**Comment by kangz12345** [comment]
global variables are placed at the beginning of the heap.
이라는 것과 관련해서 global variable(이하 gv)에 대해 질문이 있습니다.
gv라는 것이 IR 코드에서 어떤 함수에도 속하지 않은 레지스터를 말하는 것인가요?
그리고 gv가 heap의 시작부분에 놓인다는 것은 프로그램이 로딩되면서 자동으로 heap에 로드가 된 상태로 시작한다는 것인가요? 그러면 들어오는 IR 코드에서는 이러한 global variable에 접근할 때 load 등을 쓰나요..?
아니면 그냥 gv들은 heap을 사용해서 다뤄야한다는 의미인가요? 즉 프로그램 시작부분에서 global variable들은 모두 malloc으로 할당해줘야 하는 것인지 잘 모르겠습니다. 필요에 따라서 레지스터나 스택을 이용해서 gv를 다뤄도 되는지 궁금합니다.

**Comment by aqjune** [comment]
A global variable means the global variable in C (전역변수).
This IR program has a global variable x. To access to the variable, a load/store instruction should be used.
```
@x = global i32 0
define i32 @read_x() {
  %v = load i32, i32* @x
  ret i32 @x
}
```
When this IR is lowered into assembly, the naive backend will insert `x_addr = call malloc 4` at the beginning of main, and `free x_addr` at the end of main.
If the address of a global variable is not taken, you can lower it into a register or stack allocation. If you are using stack allocation instead, you should be careful so it does not raise stack overflow.

**Comment by chainpeace** [comment]
1. 스펙에 "phi" instruction은 없는 것 같은데, 다른 방법으로 loop을 구현할 수 있나요?
2. 배열타입을 쓸 수 있다고 되어있는데, "getelementptr" 없이 배열의 원소에 접근할 수 있는 방법이 궁금합니다.

**Comment by aqjune** [comment]
Note that the spec is for *assembly*, not LLVM IR.
In assembly, you can write a loop, as you write a loop in a C program.
Similarly, there is an array type in IR, but not in assembly. In assembly, getelementptr is just add of two integer values.

**Comment by aqjune** [comment]
좀 더 clarify를 하자면: 어셈블리에서 for(int i = 0; ..; i++) { } 을 쓸 수 있다는 게 아니라, *phi없이도* block들을 잘 이어붙여서 루프를 작성할 수 있다는 뜻입니다. (SSA 이야기입니다)

**Comment by tlvkfn525** [comment]
LLVM IR에서는 register를 무한정 쓸 수 있으니까, 어셈블리로 바꾸는 과정에서 register가 모자라서 stack을 사용해야 하는 경우가 생길 것 같습니다. 이런 상황은 무조건 마지막에 IR에서 어셈블리로 바꿀때 handle되는 건가요? 만약에 우리가 이 과정에 개입하고 싶으면 IR에서 alloca로 명시적으로 stack에 값을 저장해줘야 하는 건가요?

**Comment by aqjune** [comment]
> LLVM IR에서는 register를 무한정 쓸 수 있으니까, 어셈블리로 바꾸는 과정에서 register가 모자라서 stack을 사용해야 하는 경우가 생길 것 같습니다. 이런 상황은 무조건 마지막에 IR에서 어셈블리로 바꿀때 handle되는 건가요? 만약에 우리가 이 과정에 개입하고 싶으면 IR에서 alloca로 명시적으로 stack에 값을 저장해줘야 하는 건가요?

Actually this is a good question. It is called register allocation.
In general, it is the backend (IR -> assembly) that does register allocation, but if you want fine-grained control of stack usage, you'll need to annotate how to use the stack in the IR & update the backend to understand the annotation.

**Comment by micmic123** [comment]
IR에서 일반적으로
*p = malloc(8)
*q = malloc(8)
를 순차적으로 수행한 후에
p+8에 접근함으로써( Ex. load 8, p, 8 ) q에 있는 값을 가져올 수 있다는 보장이 있나요?

**Comment by aqjune** [comment]
In assembly,
```
r1 = malloc 8
r2 = malloc 8
store 8 123 r1 8
load 8, 
```
will always store 123 to the second malloc, as you said.

**Comment by micmic123** [comment]
assembly말고 IR에서는 어떤가요?

**Comment by micmic123** [comment]
```
    entry:
                %a = call i64* @malloc(i64 8)
                %b = call i64* @malloc(i64 8)
                %c = call i64* @malloc(i64 8)
                call void @free(i64* %b)
                %x = call i64* @malloc(i64 8)
                %y = call i64* @malloc(i64 8)
```
1. 위와 같은 IR 코드가 있다고 했을 때, %x + 8로 %y에 접근할 수 있다는 것이 보장되나요?
2. assembly level에서도 저렇게 중간에 땜빵(?)이 생기는 경우가 있을 수 있는데, 정말로  
> In assembly,
> 
> ```
> r1 = malloc 8
> r2 = malloc 8
> store 8 123 r1 8
> ```
> 
여기서 r2가 128이 될 수 있나요?

**Comment by aqjune** [comment]
```
r1 = malloc 8
r2 = malloc 8
store 8 123 r1 8
```
Note that there is no free between two mallocs, so it is okay.

In IR, it is a design choice, but please assume that %x + 8 cannot access %y in IR. It is undefined behavior because pointers in IR carry provenance, which is described at slides/07.UB-pointer.pdf (and the corresponding lecture video) as well.

**Comment by aqjune** [comment]
The slides don't seem to be a good reference, but the video would be better.
To summarize, for aggressive memory optimizations, it is convenient (and also correct) to assume that %x + 8 cannot access %y if %x and %y are pointing to different mallocs.
You can say that %x + ((int)%y - (int)%x) cannot access %y too, because they are originated from different mallocs.

**Comment by robinshins** [comment]
안녕하세요 조교님! 여기다 올리면 되는 질문일지 확실치 않지만
cond icmp eq %a, %a
br cond bb_true, bb_false

라는 구문을

br bb_true

로 수정했을때 메모리가 더효율적으로 쓰였다고 생각할 수 있을까요?

**Comment by aqjune** [comment]
It can use less memory because the register may not need to be spilled.

**Comment by elzino** [comment]
load나 store를 할 때 head의 움직임에 따라 cost가 발생하는데요
처음 프로그램이 실행되었을 때 head의 위치는 어디에 있는지 알 수 있을까요?

**Comment by aqjune** [comment]
The first load or store has zero cost, in terms of moving the head.

**Comment by elzino** [comment]
모듈내에서 함수가 여러번 불릴 수 있는데 이때 register value들은 보존된다고 되어 있는데요
혹시 head의 위치도 초기화가 되나요?

**Comment by aqjune** [comment]
No, head's location isn't initialized.

**Comment by kangz12345** [comment]
Is `sdiv r1 2^N` equal to `ashr r1 N`?
For example, if `r1` has the value `-1` and `N` is `1`,
then `ashr -1 1` will have the value `-1`.
In this case, what is the result of `sdiv -1 2`?

**Comment by aqjune** [comment]
sdiv -1, 2 is 0, so the optimization does not hold in general: http://volta.cs.utah.edu:8080/z/Xz4nxo
Optimizing lshr to udiv is correct: http://volta.cs.utah.edu:8080/z/wJ8XJX

**Comment by micmic123** [comment]
assembly spec에서 malloc의 size는 8의 배수라는 조건이 있었습니다.
그러면
1.  `r1 = malloc 4`와 같은 assembly 코드는 잘못된 코드인가요?
2. llvm에서 `%b = call i8* @malloc(i64 2)`처럼 써놓고 컴파일해보면 `r1 = malloc 2`로 나오던데, 그러면 주어진 simpleBackend는 해당 조건을 어긴채로 구현되어있는 것인가요? 제대로 구현했다면 에러를 던지던가 `r1 = malloc 8`로 출력돼야 맞는 듯해서요.

**Comment by aqjune** [comment]
> assembly spec에서 malloc의 size는 8의 배수라는 조건이 있었습니다.

It means the interpreter will raise runtime error.

**Comment by micmic123** [comment]
(.c)`malloc(2)` -> (.ll)`i8* @malloc(i64 2)` -> (.s) SimpleBackend: `malloc(2)` 
-> interpreter: `Runtime Error`

이렇게 SimpleBackend를 사용하면 c코드의 malloc(2)가 런타임 에러를 냅니다.
따라서 SimpleBackend가 프로젝트의 스펙을 따르도록 구현된 것이라면, **c코드 상에서 malloc(2)를 사용하면 안 된다**는 결론으로 이어지는 듯합니다.
그러나 주어진 스펙은 <assembly에서 malloc size가 8의 배수가 되야한다>는 것이므로, c코드 상에서 malloc(2) (그러므로 IR에서도 malloc(2))를 사용불가능 하다는 건 조금 이상하다는 생각이 들어 다시 질문드립니다.

c코드 상에서 malloc(2) (IR에서 malloc(2))는 잘못된 경우로 간주하면 되나요?
아니라면 저희가 SimpleBackend를 수정해서 IR에서는 malloc(2)가 가능한데 assembly에서는 그것을malloc(8)로 바뀌도록 해야하는 건가요?

**Comment by aqjune** [comment]
> c코드 상에서 malloc(2) (IR에서 malloc(2))는 잘못된 경우로 간주하면 되나요?

Yes, you can assume that the input program will never have malloc with its size non-multiple of 8.

**Comment by robinshins** [comment]
안녕하세요 조교님! 

테스트파일의 어셈블리를 보다가 이해가 가지 않는 부분이 있어 질문드립니다.

    r2 = and r1 2147483648 64
    r2 = sub 0 r2 64
    r1 = or r2 r1 64

이런 인스트럭션이 있을때

r2 = sub 0 r2 64 에서 리턴타입이 0 이라고 되어 있는것이 어떤 의미인지 모르겠습니다..!

항상 감사드립니다.

**Comment by ludaya** [comment]
안녕하세요.
spec.pdf에 Other instructions 란에
`ashr 522 2 8`
의 결과가 255가 된다는 부분을 좀 더 설명해주실수 있나요?
또 산술 연산의 결과가 64비트로 sign-extend 되는 경우는 없는지 궁금합니다.

**Comment by jinoopy** [comment]
레지스터에 대한 질문입니다.
만약 i64 숫자를 저장했다가 i32 숫자를 저장했을 경우, most significant bit으로부터 32bit에 해당하는 양[63:32]이 0으로 자동으로 초기화되나요? 아니면 garbage value가 남아있나요? 궁금합니다.

**Comment by aqjune** [comment]
@robinshins 
> r2 = sub 0 r2 64 에서 리턴타입이 0 이라고 되어 있는것이 어떤 의미인지 모르겠습니다..!
Return type of this is 64. spec.pdf will describe the syntax in detail.

@ludaya 
> ashr 522 2 8
Sorry, it was typo - this should be `ashr 511 2 8`.
The commands will always zero-extend the result.

@jinoopy 
The result of all operations are 64-bytes. Each operation will describe how the remaining bits are filled.

**Comment by robinshins** [comment]
shift right연산을 할때 저희 컴파일러는 zero extend를 하기 때문에 div로 바꿀때는 ashr 이건 lshr 이건 udiv로 바꿔줘도 상관이 없다고 봐도 될까요..?

**Comment by aqjune** [comment]
@robinshins It's a bit hard for me to answer the question; I think showing the validity of optimization is on students' side.

**Comment by tlvkfn525** [comment]
하나의 함수가 여러 번 호출되면, 두 번째로 함수가 호출되는 순간에 register의 상태는 첫 번째로 함수가 호출되고 리턴하기 직전의 register 상태와 동일한가요?
그리고 SimpleBackend 코드에서 visitfunction은 함수가 호출될 때마다 적용되는 것이 아니라 함수 하나에 대해 고유하게 한번 실행되는 코드가 맞나요?

**Comment by aqjune** [comment]
> 하나의 함수가 여러 번 호출되면, 두 번째로 함수가 호출되는 순간에 register의 상태는 첫 번째로 함수가 호출되고 리턴하기 직전의 register 상태와 동일한가요?
I think the specification says what the values of registers are after the function call. You can also check how the interpreter works as well.

> 그리고 SimpleBackend 코드에서 visitfunction은 함수가 호출될 때마다 적용되는 것이 아니라 함수 하나에 대해 고유하게 한번 실행되는 코드가 맞나요?

Well, you can write a simple program and see how it works...

**Comment by aqjune** [comment]
InstVisitor's methods will be called once per definition.

**Comment by pullup-dip-burpee** [comment]
조교님께서 올려주신 optList.md에 있는 godbolt 예시와 비슷하게 Loop Unrolling을 진행했는데 오히려 cost가 증가하였습니다 ㅠㅠ 혹시 이 경우 원인이 어떻게 될까요?

**Comment by aqjune** [comment]
In that case it depends on how your later pipeline works (e.g.  register allocation).

**Comment by pullup-dip-burpee** [comment]
조교님 loop unroll에 대해 later pipeline의 작동에 따라 코스트가 다를 수 있다고 말씀해주셨는데 잘 이해되지 않아 다시 질문드립니다... 다름이 아니라 
만약 unroll을 하기 전의 loop라는 블록의 코드가 아래와 같고
```
  .loop:
    r1 = load 8 sp 0
    r1 = add r1 1 32
    store 8 r1 sp 16
    r1 = load 8 sp 16
    r1 = icmp uge r1 8192 32
    store 8 r1 sp 24
    r1 = load 8 sp 16
    store 8 r1 sp 8
    r1 = load 8 sp 8
    store 8 r1 sp 0
    r1 = load 8 sp 24
    br r1 .exit .loop
```

unroll을 한 후의 코드가 다음과 같다면

```

  .loop:
    r1 = load 8 sp 0
    r1 = add r1 1 32
    store 8 r1 sp 16
    r1 = load 8 sp 0
    r1 = add r1 1 32
    store 8 r1 sp 24
    r1 = load 8 sp 24
    r1 = icmp uge r1 8192 32
    store 8 r1 sp 32
    r1 = load 8 sp 24
    store 8 r1 sp 8
    r1 = load 8 sp 8
    store 8 r1 sp 0
    r1 = load 8 sp 32
    br r1 .exit .loop
```

제 생각으로는 unroll을 한 이후의 코드가 icmp 이후의 부분이 호출되는 횟수가 절반이라 코스트가 훨씬 떨어져야 할 것 같은데 실제로는 cost가 오르고 있습니다... 혹시 이런 문제를 어떻게 해결할 수 있을까요?

**Comment by aqjune** [comment]
Eh.. the code seems interesting.
The value stored at (sp 16) is not used inside the loop - is it used outside the program?

**Comment by pullup-dip-burpee** [comment]
> Eh.. the code seems interesting.
> The value stored at (sp 16) is not used inside the loop - is it used outside the program?

음... 아마 그러진 않은 것 같습니다... 함수가 entry, loop, exit 세 개의 블록밖에 없어서 함수 내에서는 쓰일 일이 없고 불필요한 값인 듯합니다. 이 부분이 문제인 걸까요??

**Comment by pullup-dip-burpee** [comment]
> Eh.. the code seems interesting.
> The value stored at (sp 16) is not used inside the loop - is it used outside the program?

앗 감사합니다 덕분에 문제를 알았습니다!

**Comment by robinshins** [comment]
안녕하세요 조교님! 제가 malloc을 alloca로 바꾸어 주는 최적화를 진행하고 있는데 malloc을 분석해서 새로운 alloca를 만들어서 malloc 인스트럭션과 바꾸어주려고 하는데 replaceInstWithInst 나 replaceAllUsesWith  등을 통해서 바꾸어 주려고 하니

 void llvm::Value::doRAUW(llvm::Value*, llvm::Value::ReplaceMetadataUses): Assertion `New->getType() == getType() && "replaceAllUses of value with new value of different type!"' failed.

이와 같은 에러가 계속 뜨게 됩니다.. 아마 malloc 은 callinst 이고 alloca는 allocinst여서 타입이 달라서 이렇게 되는것이라 예상이 되는데 혹시 해결책이 있을까요..?

**Comment by aqjune** [comment]
```
void llvm::Value::doRAUW(llvm::Value*, llvm::Value::ReplaceMetadataUses): Assertion `New->getType() == getType() && "replaceAllUses of value with new value of different type!"' failed.
```
The type means IR type such as `i8*` or `i32*`.
This may happen in this case:

```
%p = alloca i32 ; %p is i32*
%q = call i8* malloc(4) ; %q is i8*
%v = load i32* %p ; It is using 4-byte load
```
If %p is replaced with %q, the last load becomes:
```
%v = load i8* %q
```
But this changes the type of %v, from i32 to i8.
To prevent this, you need to replace %p with `%q' = bitcast i8* %q to i32*`.

**Comment by tlvkfn525** [comment]
조교님, backend를 구현하는 도중에 test를 돌려보면서 conv_slot 같은 register들이 생기는 걸 확인했는데, SimpleBackend나 AssemblyEmitter 상에서는 해당 코드를 확인하지 못해 문의드립니다. 이런 게 왜 생기는지 혹시 알 수 있을까요?

**Comment by tlvkfn525** [comment]
질문에 오류가 있었네요 죄송합니다!

**Comment by lego0901** [comment]
벤치마크용 테스트 데이터를 공개해주셔서 정말 감사합니다.
혹시 아직 공개되지 않은 추가 코드나 입력 데이터가 있나요?

**Comment by aqjune** [comment]
For competition, we'll have more programs + inputs to run.

Yes, the environment will be equivalent to it.

**Comment by lego0901** [comment]
답변 정말 감사합니다. 실습 시간 전 추가로 질문하나만 더 드리겠습니다..
혹시 컴페티션에서 결과를 어떻게 measure할 계획이신가요?
[ \sum original_cost ] 와 [ \sum after_cost ] 를 비교하는건가요?
아니면 [ \sum (after_cost / original_cost) ] 를 비교하는건가요?

저희 팀에서 loop unroll을 사용하니,
작은 테스트 케이스에 대해선 소폭의 cost 증가가 나타났고 큰 입력에는 큰 캄소가 나타났습니다.
그래서 이 부분이 루프 언롤에 대한 parameter를 결정하는데 중요한 역할을 할 것 같아서 질문드립니다.

**Comment by aqjune** [comment]
To recap my answers: we're going to compare each case = (input program, input data).

**Comment by tlvkfn525** [comment]
조교님 제출을 앞두고 테스트해보는 와중에 benchmark 중 에러를 발생시키는 특이한 케이스가 있어 질문드립니다. rmq2d_sparsetable입니다. 다름이 아니라, 다음과 같이 생긴 코드를
```
  %div = udiv i64 %add, 8
  %mul = mul i64 %div, 8
```
저희가 추가한 instcombine에서 다음과 같이 변환했고
```
%__r2__ = and i64 %__r1__, -8
```
이것을 backend가 받으니 다음과 같이 변했습니다.
```
    r2 = and r1 18446744073709551608 64
```
여기서 중간에 있는 긴 수는 2^64로, 이것과 and를 하면 0이 될 수밖에 없습니다. 이 값이 malloc에 인자로 전달되기 때문에 에러가 발생한 것이었습니다.
이러한 에러를 저만 겪고 있는지 모르겠습니다만, 저희가 existing pass만 추가했을 때 버전의 컴파일러에서도 이러한 에러가 발생해 의문이 생겼습니다. -8과 and 연산을 하는 것이 동일한 효과를 불러일으킨다는 것은 논리적으로는 의심할 여지가 없어 보입니다.
backend에 아무런 pass 적용하지 않고 음수와의 and 연산을 곧바로 집어넣어봐도 같은 에러가 발생하는 것을 확인했습니다.
혹시 음수 값과의 and 연산이 허용되지 않는 것인가요? 아니면 단순한 버그인가요?
확인 부탁드립니다. 감사합니다!

**Comment by aqjune** [comment]
Hello, sorry for delay.
I think `18446744073709551608` is `2^64 - 8`: my calculator says the last 3 bits are 0 whereas all the remaining bits are 1.

**Comment by J-jaeyoung** [comment]
조교님, argument레지스터 관련 질문이 있습니다!
spec상으로 argument 레지스터를 사용할 때마다 추가코스트가 1증가한다고 이해했습니다.
```
start foo 1:
  .entry:
    r3 = add arg1 arg1 32
    ret r3
end foo
```
![image](https://user-images.githubusercontent.com/43400316/83835542-00ada100-a72c-11ea-80fd-ff0bd74a8fd6.png)

위와 같은 함수를 interpreter로 돌려보니 함수에서 cost가 2.2로 찍히던데
argument레지스터의 cost가 포함되지 않은 것 같습니다. 
확인 부탁드립니다~

**Comment by aqjune** [comment]
You're right in that TAs' intention was to increase its cost for each use of an argument. But, changing it to increase the cost per use may make function outlining less effective.
As we're in the last sprint, please follow the behavior of interpreter.

**Comment by robinshins** [comment]
안녕하세요 조교님! 

llvm intrinsic 을 사용하려고 시도중인데 저희 컴파일러로 pow 등의 instrinsic을 사용하게 되면 컴파일시에

Intrinsic has incorrect return type!
i64 (i64, i64)* @llvm.pow.i64

이런 오류가 뜨게됩니다.그냥 llvm clang으로 ll파일을 테스트 해보면 위와같은 에러가 뜨지 않는데 저희 컴파일러에 집어넣고 돌리면 뜨게됩니다.

 혹시 저희 컴파일러가 무조건 64비트로만 read write을 하기 때문일까라고 생각이 되어 관련된 모든 arg와 함수를 i64로 바꿔보기도 했는데 여전히 위와 같은 에러가 뜨는데 혹시 해결법이 있을까요..?

**Comment by aqjune** [comment]
If a intrinsic function is used, AssemblyEmitter should be updated so it lowers it into assemblies.

**Comment by kangz12345** [comment]
> You're right in that TAs' intention was to increase its cost for each use of an argument. But, changing it to increase the cost per use may make function outlining less effective.
> 
> As we're in the last sprint, please follow the behavior of interpreter.

그러면 스펙에 나온 것과 달리 아규먼트를 사용해도 추가비용이 발생하지 않는다는 말씀이신가요?

**Comment by aqjune** [comment]
It means these two instructions have the same cost:
```
r1 = add arg1 arg1 32 ; cost = 2.2
r1 = add arg1 r1 32 ; cost = 2.2
```

**Comment by robinshins** [comment]
> If a intrinsic function is used, AssemblyEmitter should be updated so it lowers it into assemblies.

혹시 어떤식으로 update를 해야할지 조금의 가이드라도 얻을 수 있을까요..? 어셈블리에 대한 이해가 부족하여 며칠동안 감을 못잡고 있습니다 ㅠ

**Comment by aqjune** [comment]
> 혹시 어떤식으로 update를 해야할지 조금의 가이드라도 얻을 수 있을까요..? 어셈블리에 대한 이해가 부족하여 며칠동안 감을 못잡고 있습니다 ㅠ
One possible solution is - in AssemblyEmitter's visitCallInst, you can check whether CI is IntrinsicInst & its intrinsic id is powi . If it is, convert it into a series of muls. InstCombineCalls.cpp will be helpful.

> 혹시 shift 연산을 mul이나 div로 바꾸어주기 위해 pow를 써야하고 그로인해서 shl 하나에서 mul이나 div 연산이 두개가 되게된다면 오히려 코스트가 늘어나는 것으로 이해할 수 있을까요..?
I don't clearly understand the situation, but if it finally converts one shl to several mul/div, then maybe yes..?
